program-test/src/programs.rs

```diff
@@ -21,76 +26,189 @@ mod spl_associated_token_account {
+/// program account.  But for the loader v3 and loader 4, upgradable programs, the relationship is a
+/// bit more complex and it could be convenient to retrieve the ELF bytes directly.
+pub fn spl_program(
+    rent: &Rent,
+    target_address: Pubkey,
+) -> Result<(ProgramElf, ProgramAccounts), ProgramLoadError> {
+    let target = SPL_PROGRAMS
+        .iter()
+        .find(|ProgramInfo { program_id, .. }| *program_id == target_address);
+
+    let target = target.ok_or(ProgramLoadError::UnknownAddress(target_address))?;
+
+    Ok((target.elf.to_vec(), program_accounts(rent, target)))
+}
+
+/// Returns accounts for all known programs, paired with their addresses.
+///
+/// Depending on the loader, a program may use one or two accounts.
+///
+/// In particular, loader v3, upgradable programs, have two accounts: one for the program and one
+/// for the program data.  In this case the program ID is the main account address.
 pub fn spl_programs(rent: &Rent) -> Vec<(Pubkey, AccountSharedData)> {
     SPL_PROGRAMS
         .iter()
-        .flat_map(|(program_id, loader_id, elf)| {
-            let mut accounts = vec![];
-            let data = if *loader_id == solana_sdk::bpf_loader_upgradeable::ID {
-                let (programdata_address, _) =
-                    Pubkey::find_program_address(&[program_id.as_ref()], loader_id);
-                let mut program_data = bincode::serialize(&UpgradeableLoaderState::ProgramData {
-                    slot: 0,
-                    upgrade_authority_address: Some(Pubkey::default()),
-                })
-                .unwrap();
-                program_data.extend_from_slice(elf);
-                accounts.push((
-                    programdata_address,
-                    AccountSharedData::from(Account {
-                        lamports: rent.minimum_balance(program_data.len()).max(1),
-                        data: program_data,
-                        owner: *loader_id,
-                        executable: false,
-                        rent_epoch: 0,
-                    }),
-                ));
-                bincode::serialize(&UpgradeableLoaderState::Program {
-                    programdata_address,
-                })
-                .unwrap()
-            } else {
-                elf.to_vec()
-            };
-            accounts.push((
-                *program_id,
-                AccountSharedData::from(Account {
-                    lamports: rent.minimum_balance(data.len()).max(1),
-                    data,
-                    owner: *loader_id,
-                    executable: true,
-                    rent_epoch: 0,
-                }),
-            ));
-            accounts.into_iter()
+        .flat_map(|info| {
+            let accounts = program_accounts(rent, info);
+            Vec::<_>::from(accounts).into_iter()
         })
         .collect()
 }
+
+fn program_accounts(rent: &Rent, info: &ProgramInfo) -> ProgramAccounts {
```

Comment:
Nitty, but why not make these internal methods on `ProgramAccounts`? Something like:
```
fn from(rent: &Rent, info: &ProgramInfo) -> Self { .. }
fn loader_v2(rent: &Rent, address: Pubkey, elf: &[u8]) -> Self { .. }
fn loader_v3(rent: &Rent, main_address: Pubkey, elf: &[u8]) -> { .. }
```

Reply:
We are talking about a module that literally has 5 functions in it.
And most are there to construct the `ProgramAccounts`.
I am not sure renaming `program_accounts_loader_v2` to `ProgramAccounts::loader_v2` changes anything in terms of readability.
It will just add a level of indentation to the function implementation.